### PR TITLE
kubelet: Add a TopologyManager policy option: max-allowable-numa-nodes

### LINKED ISSUE #124144: Make maxAllowableNUMANodes in the kubelet's TopologyManager configurable
### What would you like to be added?

Add a config option to the `kubelet` that allows us to configure the value of `maxAllowableNUMANodes` in the TopologyManager:
https://github.com/kubernetes/kubernetes/blob/79c61d5f0305c566b642be978b1e4837870db257/pkg/kubelet/cm/topologymanager/topology_manager.go#L40

### Why is this needed?

The current hard-coded value of 8 was added as a stop-gap 4 years ago, with a plan to improve the topology manager to remove the state explosion that occurs when trying to enumerate the possible NUMA affinities and generating their hints (which unfortunately never happened).

The long term solution should be to refactor this code to avoid the state explosion altogether. In the meantime, a configurable option should be sufficient to allow users to increase the limit when appropriate.

### PR DESCRIPTION
**Pull request by cyclinder**
Add a TopologyManager policy option for kubelet, which can make maxAllowableNUMANodes in the kubelet's TopologyManager configurable.

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Add a TopologyManager policy option for kubelet, which can make maxAllowableNUMANodes in the kubelet's TopologyManager configurable.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/124144

#### Does this PR introduce a user-facing change?

```release-note
Add a TopologyManager policy option: max-allowable-numa-nodes to configures maxAllowableNUMANodes for kubelet.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/pull/4624
```

### CODE REVIEW

File: cmd/kubelet/app/options/options.go [hunk 1]
```diff
@@ -90,6 +90,16 @@ type KubeletFlags struct {
 	// additional configs to overwrite what is provided by default and in the KubeletConfigFile flag
 	KubeletDropinConfigDirectory string
 
+	// maxAllowableNUMANodes specifies the maximum number of NUMA Nodes that
```
Comment (kannon92) [inline cmd/kubelet/app/options/options.go:93]: This should be moved to the kubelet config I think. We are aiming to avoid adding parameters to the kubelet CLI. I'd add it both in the staging directory and pkg/kubelet/apis/config.

**Review by kannon92 — CHANGES_REQUESTED** [review]
https://github.com/kubernetes/kubernetes/pull/124148/files#r1547741506

File: pkg/kubelet/cm/topologymanager/topology_manager_test.go [hunk 2]
```diff
@@ -105,7 +105,7 @@ func TestNewManager(t *testing.T) {
 			description:    "more than 8 NUMA nodes",
 			policyName:     "best-effort",
 			expectedPolicy: "best-effort",
-			expectedError:  fmt.Errorf("unsupported on machines with more than %v NUMA Nodes", maxAllowableNUMANodes),
+			expectedError:  fmt.Errorf("unsupported on machines with more than %v NUMA Nodes", 8),
```
Comment (kannon92) [inline pkg/kubelet/cm/topologymanager/topology_manager_test.go:108]: Let's store 8 as a variable and reference it in the tests.

File: cmd/kubelet/app/options/options.go (hunk 1 above)
Comment (cyclinder) [reply cmd/kubelet/app/options/options.go:93]: thanks for the review. 

> We are aiming to avoid adding parameters to the kubelet CLI.

yes, makes sense, I'll add it later.

Comment (klueska) [reply cmd/kubelet/app/options/options.go:93]: I also don't think this should be a top-level flag, but rather a topology-manager policy option. See the following for the PR that added policy options, including the very first option: https://github.com/kubernetes/kubernetes/pull/112914

Comment (ffromani) [reply cmd/kubelet/app/options/options.go:93]: A policy option would indeed let us plug in the existing framework without new API reviews. Perhaps this is the way to go for all the new flags and features for topology manager?

File: cmd/kubelet/app/options/options.go [hunk 3]
```diff
@@ -90,6 +90,16 @@ type KubeletFlags struct {
...
+	// the TopologyManager supports on the underlying machine.
+	//
+	// At present, having more than this number of NUMA Nodes will result in a
+	// state explosion when trying to enumerate possible NUMAAffinity masks and
+	// generate hints for them. As such, if more NUMA Nodes than this are
+	// present on a machine and the TopologyManager is enabled, an error will
+	// be returned and the TopologyManager will not be loaded.
+	MaxAllowableNUMANodes int
```
Comment (ffromani) [inline cmd/kubelet/app/options/options.go:101]: For top-level internal flags this name is mostly OK - it should however perhaps have the `TopologyManager` prefix, let's review how other related flags are named. OTOH if you move towards topology manager policy (which I'd be generally in favor of) you can keep this name internally.

File: pkg/kubelet/cm/topologymanager/topology_manager.go [hunk 4]
```diff
@@ -29,15 +29,6 @@ import (
...
 const (
-	// maxAllowableNUMANodes specifies the maximum number of NUMA Nodes that
-	// the TopologyManager supports on the underlying machine.
-	//
-	// At present, having more than this number of NUMA Nodes will result in a
-	// state explosion when trying to enumerate possible NUMAAffinity masks and
-	// generate hints for them. As such, if more NUMA Nodes than this are
-	// present on a machine and the TopologyManager is enabled, an error will
-	// be returned and the TopologyManager will not be loaded.
-	maxAllowableNUMANodes = 8
```
Comment (ffromani) [inline pkg/kubelet/cm/topologymanager/topology_manager.go:40]: I'd keep this constant but renamed like `DefaultMaxAllowableNUMANodes` - if user doesn't specify a new value, this is the recommended default. The user should probably not be enabled to set less (what's the point?) but if they set more, they should IMO get a warning both in the logs and from the docs

File: cmd/kubelet/app/options/options.go (hunk 1 above)
Comment (swatisehgal) [reply cmd/kubelet/app/options/options.go:93]: >  I also don't think this should be a top-level flag, but rather a topology-manager policy option. See the following for the PR that added policy options, including the very first option: 
https://github.com/kubernetes/kubernetes/pull/112914.

Interesting, I would have thought that this seems more apt as a top level kubelet flag behind a feature gate given its nature and relevance across all policies and policy options. If the value is not specified and feature gate not enabled, we can default to the old value of 8 and preserve backward compatibility with minimal code changes.

We would have to go through graduation process irrespective of its implementation as a top level flag or a policy option. What is the rationale behind proposing this as a policy option?

**Review by swatisehgal — COMMENTED** [review]
We would need a KEP for this work so please ensure that this is captured in the Node 1.31 planning document: https://docs.google.com/document/d/1U10J0WwgWXkdYrqWGGvO8iH2HKeerQAlygnqgDgWv4E/edit.

File: cmd/kubelet/app/options/options.go (hunk 1 above)
Comment (cyclinder) [reply cmd/kubelet/app/options/options.go:93]: So what's the conclusion here? We only need a policy option or a feature-gate. I prefer to have a policy option here.

File: cmd/kubelet/app/options/options.go (hunk 3 above)
Comment (cyclinder) [reply cmd/kubelet/app/options/options.go:101]: yes, If we consider it as a policy option, It should have the `TopologyManager` prefix.

File: pkg/kubelet/cm/topologymanager/topology_manager.go (hunk 4 above)
Comment (cyclinder) [reply pkg/kubelet/cm/topologymanager/topology_manager.go:40]: That make sense, thanks for the comment.

File: cmd/kubelet/app/options/options.go (hunk 3 above)
Comment (klueska) [reply cmd/kubelet/app/options/options.go:101]: As a policy option it should *not* have the `TopologyManager` prefix.

File: cmd/kubelet/app/options/options.go (hunk 1 above)
Comment (klueska) [reply cmd/kubelet/app/options/options.go:93]: I still believe it should be a policy option. It scopes it beneath the TopologyManager better than having a top level flag.

File: cmd/kubelet/app/options/options.go (hunk 3 above)
Comment (cyclinder) [reply cmd/kubelet/app/options/options.go]: Sorry for my bad words, yes, we should keep `MaxAllowableNUMANodes`  name internally.

File: cmd/kubelet/app/options/options.go (hunk 1 above)
Comment (cyclinder) [reply cmd/kubelet/app/options/options.go]: Based on your comments, I've made some updates, Could you retake a review? thanks!

File: pkg/kubelet/cm/topologymanager/topology_manager.go (hunk 4 above)
Comment (cyclinder) [reply pkg/kubelet/cm/topologymanager/topology_manager.go:40]: > but if they set more, they should IMO get a warning both in the logs and from the docs

Can you tell me where this warning or log or docs would be better if they set more? thanks!

File: pkg/kubelet/cm/topologymanager/policy_options.go [hunk 5]
```diff
@@ -27,12 +27,14 @@ import (
...
 const (
 	PreferClosestNUMANodes string = "prefer-closest-numa-nodes"
+	MaxAllowableNUMANodes  string = "max-allowable-numa-nodes"
 )
 
 var (
 	alphaOptions = sets.New[string]()
 	betaOptions  = sets.New[string](
 		PreferClosestNUMANodes,
+		MaxAllowableNUMANodes,
```
Comment (PiotrProkop) [inline pkg/kubelet/cm/topologymanager/policy_options.go:37]: The new option should start as `alpha` option and graduate independently.

Comment (cyclinder) [reply pkg/kubelet/cm/topologymanager/policy_options.go:37]: Thanks! Done.

**Comment by cyclinder** [comment]
Hi @PiotrProkop @ffromani Could you please take a look?

**Comment by PiotrProkop** [comment]
The changes looks good to me. I just think we need KEP for this effort(should be pretty straightforward, I can help with the KEP if needed) and when it will be approved we can track this change for k8s 1.31 😄

**Comment by cyclinder** [comment]
> require some extra process as it stands today

Is the extra process a KEP or something else? I am willing to submit a KEP for this 😄

**Comment by ffromani** [comment]
> > require some extra process as it stands today
> 
> Is the extra process a KEP or something else? I am willing to submit a KEP for this 😄

Regarding the extra process I was thinking to the API review for the flag addition. AFAIK the flag addition won't need a Feature Gate nor a KEP, at least not in the general case. If so, it will emerge during the API review I think.
For the policy option, it will require a KEP as the process stands today AFAIK.

**Comment by klueska** [comment]
I still think this should be TopologyManager policy option, but we should have it start as an already-on beta feature according to the new policy. @cyclinder do you have the tie / experience to put together a KEP for this? I'm happy to help review.

**Comment by klueska** [comment]
Here is an example for adding a similar policy option to the CPU manager. I would imagine this KEP to look similar:
https://github.com/kubernetes/enhancements/pull/3334

**Comment by cyclinder** [comment]
Thanks @klueska! I'm interested in this and I'll spend the time understanding this.

**Review by ffromani — COMMENTED** [review]
it seems we need some updates to reflect the KEP changes

**Comment by ffromani** [comment]
In addition, if we start as beta, we should do some e2e changes. I'd start with a simple modularization: change the existing e2e tests to run once with default settings, and again the same tests with `maxAllowableNUMANodes = 2*default`. This is the most basic verification that the changed setting is not breaking stuff.
We are unlikely to ever hardware powerful enough to exercise > 8 NUMA Nodes (or just 4...), so this very basic validation is likely the best we can do in the short/mid term

File: cmd/kubelet/app/options/options.go [hunk 6]
```diff
@@ -501,7 +501,7 @@ func AddKubeletConfigFlags(mainfs *pflag.FlagSet, c *kubeletconfig.KubeletConfig
...
 	fs.StringVar(&c.ReservedSystemCPUs, "reserved-cpus", c.ReservedSystemCPUs, "A comma-separated list of CPUs or CPU ranges that are reserved for system and kubernetes usage. This specific list will supersede cpu counts in --system-reserved and --kube-reserved.")
 	fs.StringVar(&c.TopologyManagerScope, "topology-manager-scope", c.TopologyManagerScope, "Scope to which topology hints applied. Topology Manager collects hints from Hint Providers and applies them to defined scope to ensure the pod admission. Possible values: 'container', 'pod'.")
-	fs.Var(cliflag.NewMapStringStringNoSplit(&c.TopologyManagerPolicyOptions), "topology-manager-policy-options", "A set of key=value Topology Manager policy options to use, to fine tune their behaviour. If not supplied, keep the default behaviour.")
+	fs.Var(cliflag.NewMapStringString(&c.TopologyManagerPolicyOptions), "topology-manager-policy-options", "A set of key=value Topology Manager policy options to use, to fine tune their behaviour. If not supplied, keep the default behaviour.")
```
[the remark this answers is not in the document]
Comment (cyclinder) [reply cmd/kubelet/app/options/options.go:504]: `TopologyManagerPolicyOptions` can set more than one option, Use `NewMapStringString` to pass multiple sets at the same time, commas separate them. see https://github.com/kubernetes/kubernetes/blob/88313a445174e21ed326f40802429b854e5be9ba/staging/src/k8s.io/component-base/cli/flag/map_string_string.go#L76-L90

And I've seen other similar kubelet flags that also use `NewMapStringString`, see https://github.com/kubernetes/kubernetes/blob/88313a445174e21ed326f40802429b854e5be9ba/cmd/kubelet/app/options/options.go#L497

please let me know If I'm wrong.

File: pkg/kubelet/cm/topologymanager/policy_options.go [hunk 7]
```diff
@@ -71,11 +75,23 @@ func NewPolicyOptions(policyOptions map[string]string) (PolicyOptions, error) {
 				return opts, fmt.Errorf("bad value for option %q: %w", name, err)
 			}
 			opts.PreferClosestNUMA = optValue
+		case MaxAllowableNUMANodes:
+			optValue, err := strconv.Atoi(value)
+			if err != nil {
+				return opts, fmt.Errorf("bad value for option %q: %w", name, err)
+			}
```
[the remark this answers is not in the document]
Comment (cyclinder) [reply pkg/kubelet/cm/topologymanager/policy_options.go:82]: yes, updated.

File: cmd/kubelet/app/options/options.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ffromani) [reply cmd/kubelet/app/options/options.go:504]: it seems a correct change, thanks for elaborating.

**Comment by cyclinder** [comment]
>  if we start as beta, we should do some e2e changes. I'd start with a simple modularization: change the existing e2e tests to run once with default settings, and again the same tests with maxAllowableNUMANodes = 2*default

It's ok to me https://github.com/kubernetes/kubernetes/blob/88313a445174e21ed326f40802429b854e5be9ba/test/e2e_node/topology_manager_test.go#L963 is that e2e test? we just run the test again with the update maxAllowableNUMANodes ( 2*default, 16) of the kubelet config.

**Comment by ffromani** [comment]
> > if we start as beta, we should do some e2e changes. I'd start with a simple modularization: change the existing e2e tests to run once with default settings, and again the same tests with maxAllowableNUMANodes = 2*default
> 
> It's ok to me
> 
> https://github.com/kubernetes/kubernetes/blob/88313a445174e21ed326f40802429b854e5be9ba/test/e2e_node/topology_manager_test.go#L963
> is that e2e test? I'm not sure, we just run the test again with the update maxAllowableNUMANodes ( 2*default, 16) of the kubelet config.

Yes, that's it. It is possible we "just" need another Context which sets `maxAllowableNUMANodes =  2 * default` and runs the same tests. Let's make sure we take the chance to add tags and labels so it's easy to disable tests before run (ginkgo skip) or run a subset of them.

File: test/e2e_node/topology_manager_test.go [hunk 8]
```diff
@@ -960,6 +964,10 @@ var _ = SIGDescribe("Topology Manager", framework.WithSerial(), feature.Topology
 	f.NamespacePodSecurityLevel = admissionapi.LevelPrivileged
 
 	ginkgo.Context("With kubeconfig updated to static CPU Manager policy run the Topology Manager tests", func() {
-		runTopologyManagerTests(f)
+		runTopologyManagerTests(f, nil)
+	})
+	ginkgo.Context("With kubeconfig's topologyOptions updated to static CPU Manager policy run the Topology Manager tests", ginkgo.Label("MaxAllowableNUMANodes"), func() {
+		allowableNUMANodes := strconv.Itoa(topologymanager.DefaultMaxAllowableNUMANodes * 2)
```
Comment (ffromani) [inline test/e2e_node/topology_manager_test.go:970]: we need to think a tiny bit more about the naming but yes, this is what I meant and the change generally LGTM. Thanks!

**Comment by cyclinder** [comment]
Thanks for the reminder. I've added another e2e test for it, any comments are appreciated!

File: test/e2e_node/topology_manager_test.go (hunk 8 above)
Comment (cyclinder) [reply test/e2e_node/topology_manager_test.go:970]: how about the doubleDefaultMaxAllowableNUMANodes? Do you have a better suggestion?

**Comment by ffromani** [comment]
> /test pull-kubernetes-node-e2e-containerd-serial-ec2 /test pull-kubernetes-node-kubelet-serial-containerd /test pull-kubernetes-node-kubelet-serial-cpu-manager /test pull-kubernetes-node-kubelet-serial-topology-manager

these are pretty expensive lanes, let's trigger them when the fast lanes (linter-hints, lint...) are all green

**Comment by cyclinder** [comment]
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/124148/pull-kubernetes-verify-lint/1808798941089632256

```
ERROR: pkg/kubelet/cm/topologymanager/policy_options.go:90:5: unstructured logging function "Warningf" should not be used (logcheck)
ERROR:                          klog.Warningf("WARNING: the value of %v is more than the default recommended value: %v", name, DefaultMaxAllowableNUMANodes)
ERROR:   
```

It seems we have to use structured logging function, so I would use `klog.Infos` here instead of `klog.Warningf`.

**Comment by cyclinder** [comment]
okay, Now the fast lanes are all green, let's trigger them!

/test pull-kubernetes-node-e2e-containerd-serial-ec2
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

**Comment by ffromani** [comment]
thansk @cyclinder 
unfortunately it seems the topology manager lane is broken:
```
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
>                              START TEST                                >
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
Start Test Suite on Host n2d-standard-32-cos-beta-113-18244-85-49-b4bc93b3
I0705 08:56:36.498619    1735 factory.go:193] Registered Plugin "containerd"
  I0705 08:56:36.524485    1735 mount_linux.go:334] Detected umount with safe 'not mounted' behavior
  I0705 08:56:36.525766    1735 mount_linux.go:334] Detected umount with safe 'not mounted' behavior
  W0705 08:56:36.541926    1735 test_context.go:538] Unable to find in-cluster config, using default host : https://127.0.0.1:6443
  I0705 08:56:36.541963    1735 test_context.go:553] Tolerating taints "node-role.kubernetes.io/control-plane" when considering if nodes are ready
  I0705 08:56:36.541970 1735 test_context.go:561] The --provider flag is not set. Continuing as if --provider=skeleton had been used.
  I0705 08:56:36.542112    1735 feature_gate.go:385] feature gates: {map[]}
  I0705 08:56:36.542181    1735 feature_gate.go:385] feature gates: {map[]}
  I0705 08:56:36.550415    1735 mount_linux.go:334] Detected umount with safe 'not mounted' behavior
  I0705 08:56:36.551745    1735 mount_linux.go:334] Detected umount with safe 'not mounted' behavior
Running Suite: E2eNode Suite - /tmp/node-e2e-20240705T085624
============================================================
Random Seed: 1720169796 - will randomize all specs

Will run 0 of 592 specs
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
[ReportAfterSuite] Kubernetes e2e JUnit report
k8s.io/kubernetes/test/e2e/framework/test_context.go:612
[ReportAfterSuite] PASSED [0.003 seconds]
------------------------------

Ran 0 of 592 Specs in 0.008 seconds
SUCCESS! -- 0 Passed | 0 Failed | 0 Pending | 592 Skipped
```

**Comment by cyclinder** [comment]
It seem the topology manager lane is passed, ah,  and https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/124148/pull-kubernetes-node-e2e-containerd-serial-ec2/1809146189652692992 is don't related to these changes, right?

/test pull-kubernetes-node-e2e-containerd-serial-ec2

**Comment by ffromani** [comment]
> It seem the topology manager lane is passed, ah, and https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/124148/pull-kubernetes-node-e2e-containerd-serial-ec2/1809146189652692992 is doesn't related to these changes, right?
> 
> /test pull-kubernetes-node-e2e-containerd-serial-ec2

it seems to be passing _because it skipped all the tests_, which would be scary, a problem but totally not related to this PR

**Comment by ffromani** [comment]
> > it seems to be passing _because it skipped all the tests_, which would be scary, a problem but totally not related to this PR
> 
> ah, why are all the tests being skipped？

this is what we need to figure out now. Again, not related or caused by your PR, but it's getting in the way of the verification.

**Comment by ffromani** [comment]
@cyclinder no worries, we'll investigate on a separate PR so this one won't be blocked. Could you please retitle this PR and fits its release note text (first message) to convey we implemented using a topology manager option?
I'll do a final review ASAP.

File: test/e2e_node/topology_manager_test.go [hunk 9]
```diff
@@ -203,13 +203,17 @@ func findNUMANodeWithoutSRIOVDevices(configMap *v1.ConfigMap, numaNodes int) (in
 	return findNUMANodeWithoutSRIOVDevicesFromSysfs(numaNodes)
 }
 
-func configureTopologyManagerInKubelet(oldCfg *kubeletconfig.KubeletConfiguration, policy, scope string, configMap *v1.ConfigMap, numaNodes int) (*kubeletconfig.KubeletConfiguration, string) {
+func configureTopologyManagerInKubelet(oldCfg *kubeletconfig.KubeletConfiguration, policy, scope string, topologyOptions map[string]string, configMap *v1.ConfigMap, numaNodes int) (*kubeletconfig.KubeletConfiguration, string) {
 	// Configure Topology Manager in Kubelet with policy.
 	newCfg := oldCfg.DeepCopy()
 	if newCfg.FeatureGates == nil {
 		newCfg.FeatureGates = make(map[string]bool)
```
Comment (ffromani) [inline test/e2e_node/topology_manager_test.go:210]: we need to enable the TopologyManager Alpha-level policy options to make the current code work. But didn't we decide in the KEP to start from beta?

**Comment by cyclinder** [comment]
> no worries, we'll investigate on a separate PR so this one won't be blocked. Could you please retitle this PR and fits its release note text (first message) to convey we implemented using a topology manager option?
I'll do a final review ASAP.

Sorry for the delay. let me update the texts now.

File: pkg/kubelet/cm/topologymanager/policy_options.go (hunk 5 above)
Comment (klueska) [reply pkg/kubelet/cm/topologymanager/policy_options.go:36]: Yes, this needs to change. The new option should be put directly into beta.

File: pkg/kubelet/cm/topologymanager/policy_options.go [hunk 10]
```diff
@@ -71,11 +76,31 @@ func NewPolicyOptions(policyOptions map[string]string) (PolicyOptions, error) {
 				return opts, fmt.Errorf("bad value for option %q: %w", name, err)
 			}
 			opts.PreferClosestNUMA = optValue
+		case MaxAllowableNUMANodes:
+			optValue, err := strconv.Atoi(value)
+			if err != nil {
+				return opts, fmt.Errorf("bad value for option %q: %w", name, err)
```
Comment (klueska) [inline pkg/kubelet/cm/topologymanager/policy_options.go:82]:
```suggestion
				return opts, fmt.Errorf("unable to convert policy option to integer %q: %w", name, err)
```

File: pkg/kubelet/cm/topologymanager/policy_options.go [hunk 11]
```diff
@@ -71,11 +76,31 @@ func NewPolicyOptions(policyOptions map[string]string) (PolicyOptions, error) {
...
 			// but we keep it as further safety.
 			return opts, fmt.Errorf("unsupported topologymanager option: %q (%s)", name, value)
 		}
 	}
+
+	// if user doesn't specify a new value, make defaultMaxAllowableNUMANodes
+	// is the recommended default.
+	if opts.MaxAllowableNUMANodes == 0 {
+		opts.MaxAllowableNUMANodes = DefaultMaxAllowableNUMANodes
+	}
```
Comment (klueska) [inline pkg/kubelet/cm/topologymanager/policy_options.go:104]: Can we instead move this to the top and do it unconditionally?, i.e.
```
// Set MaxAllowableNUMANodes to the default. This will be overwritten
// if the user has specified a policy option for MaxAllowableNUMANodes.
opts.MaxAllowableNUMANodes = DefaultMaxAllowableNUMANodes
```

File: test/e2e_node/topology_manager_test.go (hunk 9 above)
Comment (ffromani) [reply test/e2e_node/topology_manager_test.go:210]: either works, but the best fix is to move the new option to beta.

File: pkg/kubelet/cm/topologymanager/topology_manager.go [hunk 12]
```diff
@@ -29,15 +29,15 @@ import (
...
+	// defaultMaxAllowableNUMANodes specifies the maximum number of NUMA Nodes that
 	// the TopologyManager supports on the underlying machine.
 	//
 	// At present, having more than this number of NUMA Nodes will result in a
 	// state explosion when trying to enumerate possible NUMAAffinity masks and
 	// generate hints for them. As such, if more NUMA Nodes than this are
 	// present on a machine and the TopologyManager is enabled, an error will
 	// be returned and the TopologyManager will not be loaded.
-	maxAllowableNUMANodes = 8
+	DefaultMaxAllowableNUMANodes = 8
```
Comment (klueska) [inline pkg/kubelet/cm/topologymanager/topology_manager.go:40]: Is there a reason to use a capital `D` to export this variable outside of the package?

File: test/e2e_node/topology_manager_test.go (hunk 9 above)
Comment (klueska) [reply test/e2e_node/topology_manager_test.go:210]: We shouldn't have to, but you still have the option listed as an alpha feature here:
https://github.com/kubernetes/kubernetes/pull/124148/files#diff-441ff1f99d4d01db0b4d2ea6b26faee508b87af24e6dbae93229126cd4198cbaR36

Comment (cyclinder) [reply test/e2e_node/topology_manager_test.go:210]: Isn't this new option already in beta? see https://github.com/kubernetes/kubernetes/pull/124148#discussion_r1558954149

File: pkg/kubelet/cm/topologymanager/topology_manager.go (hunk 12 above)
Comment (cyclinder) [reply pkg/kubelet/cm/topologymanager/topology_manager.go:40]: thanks! you are right, I was using a lowercase `d`, and then I referenced it in the test file, and then my vs-code reported an error and I thought I needed an uppercase `D`, and now that I've changed it back, everything is fine.

File: test/e2e/framework/test_context.go [hunk 13]
```diff
@@ -399,7 +399,7 @@ func CreateGinkgoConfig() (types.SuiteConfig, types.ReporterConfig) {
 	// Randomize specs as well as suites
 	suiteConfig.RandomizeAllSpecs = true
 	// Disable skipped tests unless they are explicitly requested.
-	if len(suiteConfig.FocusStrings) == 0 && len(suiteConfig.SkipStrings) == 0 {
+	if len(suiteConfig.FocusStrings) == 0 && len(suiteConfig.SkipStrings) == 0 && suiteConfig.LabelFilter == "" {
```
Comment (cyclinder) [inline test/e2e/framework/test_context.go:402]: waiting for https://github.com/kubernetes/kubernetes/pull/125966 merged, just for testing, revert it when https://github.com/kubernetes/kubernetes/pull/125966 merged.

File: pkg/kubelet/cm/topologymanager/policy_options.go (hunk 5 above)
Comment (ffromani) [reply pkg/kubelet/cm/topologymanager/policy_options.go:35]: it seems this is still in the `alpha` options in the last upload?

File: pkg/kubelet/cm/topologymanager/policy_options.go (hunk 10 above)
Comment (ffromani) [reply pkg/kubelet/cm/topologymanager/policy_options.go:81]: still pending?

File: pkg/kubelet/cm/topologymanager/policy_options.go [hunk 14]
```diff
@@ -71,11 +75,23 @@ func NewPolicyOptions(policyOptions map[string]string) (PolicyOptions, error) {
...
 			// but we keep it as further safety.
 			return opts, fmt.Errorf("unsupported topologymanager option: %q (%s)", name, value)
 		}
 	}
+
+	// if user doesn't specify a new value, make defaultMaxAllowableNUMANodes
+	// is the recommended default.
+	if opts.MaxAllowableNUMANodes == 0 {
+		opts.MaxAllowableNUMANodes = defaultMaxAllowableNUMANodes
+	}
```
Comment (ffromani) [inline pkg/kubelet/cm/topologymanager/policy_options.go:95]: let's move this initialization at the beginning, around line 65:
```
opts := PolicyOptions{
    MaxAllowableNUMANodes: defaultMaxAllowableNUMANodes,
}
```

Comment (klueska) [reply pkg/kubelet/cm/topologymanager/policy_options.go:95]: Hmm, I had left the same comment but it is lost now somehow...

**Comment by ffromani** [comment]
> Yes, My home computer did not fetch the latest changes, resulting in the loss of today's changes. Now I have restored them.

No problem! thanks for the quick update. Will review again soon. We should be close to merge.

**Comment by cyclinder** [comment]
> Just checking in as we approach code freeze at 02:00 UTC Wednesday 24th July 2024 / 19:00 PDT Tuesday 23rd July 2024.

It seems all the tests have passed, Can this PR be merged before code freeze? @ffromani @klueska

File: pkg/kubelet/cm/topologymanager/policy_options_test.go [hunk 15]
```diff
@@ -112,7 +145,7 @@ func TestNewTopologyManagerOptions(t *testing.T) {
 	}
 
 	betaOptions.Insert(fancyBetaOption)
-	alphaOptions = sets.New[string](fancyAlphaOption)
+	alphaOptions.Insert(fancyAlphaOption)
```
Comment (ffromani) [inline pkg/kubelet/cm/topologymanager/policy_options_test.go:148]: nit: unnecessary change, but please do not resubmit just to fix this

**Review by ffromani — COMMENTED** [review]
/lgtm

File: pkg/kubelet/cm/topologymanager/policy_options_test.go (hunk 15 above)
Comment (cyclinder) [reply pkg/kubelet/cm/topologymanager/policy_options_test.go:148]: Is it necessary to fix it later?

Comment (ffromani) [reply pkg/kubelet/cm/topologymanager/policy_options_test.go:148]: IMO we can keep it like this.

Comment (ffromani) [reply pkg/kubelet/cm/topologymanager/policy_options_test.go:148]: Let's check the tests once this PR is merged, is going to be a lower-priority test improvement

**Review by mrunalp — APPROVED** [review]
